File: nextGreaterElem.py
#Leetcode : pass
#TC- O(n) SP-O(1)
#1)create result array with -1
#2)push each elem index into stack
#3)loop over the array twice as it's circular array
#4)check while stack !=Null and stack_top < curr
##4.1)store curr elem

class Solution:
    def nextGreaterElements(self, nums):
        #1
        result = [-1]*len(nums)
        stack =[]
        #3
        for x in range(len(nums)*2):
            i = x%len(nums)
            #4
            while len(stack) != 0 and nums[i] > nums[stack[-1]]:
                #4.1
                result[stack.pop()] = nums[i]
            #2    
            stack.append(i)
        return result

obj = Solution()
print(dailyTemperatures([1,2,1]))

File: NextGreaterElem.java
// tc = O(n)
// sc = O(n)

public class NextGreaterElem {

}

class Solution {
    public int[] nextGreaterElements(int[] nums) {
        // if(nums == null && nums.length == 0) return -1;
        int n = nums.length;
        int[] result = new int[n];
        Stack<Integer> st = new Stack<>();
        Arrays.fill(result, -1);
        for (int i = 0; i < 2 * n; i++) {
            while (!st.isEmpty() && nums[st.peek()] < nums[i % n]) {
                int popped = st.pop();
                result[popped] = nums[i % n];
            }
            if (i < n) {
                st.push(i);
            }
        }
        return result;

    }
}